### PR TITLE
Issue 885/survey response singular 

### LINKED ISSUE #885: View column modal choice: Survey response (singular)
## Description
A user should be able to add a column to a view which contains the response to a single survey. This column choice creates different types of columns based on the type of the survey question.

## Relevant Job Stories
None

## Prerequisites
Don't start this until #875 is done

## Requirements
- [ ] Picking this choice should progress the modal to the config state
- [ ] In the config state, the user must first pick a survey
- [ ] In the next state, the user must pick a question
- [ ] The next state varies depending on the question type
  - [ ] For free text questions, just display a text (no options) and allow the user to finish, to create a `survey_response` type column
  - [ ] For questions with options, show a list with all the options
    - [ ] If user picks an option, create single `survey_option` column
    - [ ] If user picks special option "All options (single column)", create `survey_options` column
    - [ ] If user picks special option "All options (separate columns)", create multiple `survey_option` columns

## Design specifications
Uses the unified design language of the column modal framework.

* **TITLE**: Survey response
* **DESCRIPTION**: Include response to a question from any survey
* **ACTIONS**: Configure

## Open questions
None

### PR DESCRIPTION
**Pull request by rebecarubio**
## Description
This PR allows the user to select survey response from survey questions and chose different options to get the response shown in the view

## Screenshots
a-1
a2
a-3

## Changes
* Adds survey_option column type
* Adds survey_response modal choice 
* Adds the survey_response modal choice configuration

## Related issues
Resolves #885

### CODE REVIEW

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+interface SurveyResponseConfigProps {
+  onOutputConfigured: (columns: SelectedViewColumn[]) => void;
+}
+
+export enum SURVEY_QUESTION_OPTIONS {
+  ALL_OPTIONS = 'AllSingle',
+  ALL_OPTIONS_SEPARATED = 'AllOptionsSeparated',
+  ONE_OPTION = 'oneOption',
+}
```
Comment (ziggabyte) [inline src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:31]: consistent casing in the strings

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        const optionTitle = selectedQuestion.question.options?.find(
+          (option) => option.id === (selectedOption as unknown as number)
+        );
+        if (optionTitle === undefined) {
+          return [];
+        }
+        return [
+          {
+            config: {
+              option_id: selectedOption as unknown as number,
```
Comment (ziggabyte) [inline src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:105]: i get really curious about this type casting and assume there is a story here?

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+                </ListSubheader>
+                {selectedQuestion?.question.options?.map((option) => (
+                  <MenuItem key={option.id} value={option.id}>
+                    {option.text}
+                  </MenuItem>
+                ))}
+              </TextField>
+            ) : (
+              ''
+            )}
```
Comment (ziggabyte) [inline src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:218]: I am curious about these empty strings in the ternaries - do they do something different from the conditional render using "&&" ?

File: src/features/views/components/ViewColumnDialog/SurveyResponsePluralConfig.tsx [hunk 4]
```diff
@@ -76,32 +76,35 @@ const SurveyResponsesConfig = ({
...
+                  <TextField
+                    {...params}
+                    inputProps={{
+                      ...params.inputProps,
+                    }}
+                    label={intl.formatMessage({
+                      id: 'misc.views.columnDialog.choices.surveyResponses.questionField',
+                    })}
+                    variant="standard"
+                  ></TextField>
```
Comment (ziggabyte) [inline src/features/views/components/ViewColumnDialog/SurveyResponsePluralConfig.tsx:101]: Let this be self-closing maybe?

**Review by ziggabyte — COMMENTED** [review]
Great job tackling the boss that was too big on the first day! When I play with it I find nothing unexpected :D 🐎  only one thing that has more to do with surveys - it gets a bit weird when there are no choices in a choices question. But i made a note of in slack that I think we should fix when working on surveys.

I made some comments that are mostly from curiosity of knowing the process, and little style things :)

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx (hunk 2 above)
Comment (richardolsson) [reply src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:105]: Yes, this is problematic. The reason for this cast is that `option_id` expects a `number`, but according to the function argument list, `selectedOption` is either a `string` or a `SURVEY_QUESTION_OPTION`. My guess is that @rebecarubio has tried to cast it as `number` directly, but there is no overlap between the `string | SURVEY_QUESTION_OPTION` union and `number`, so Typescript has recommended that it be cast "through" `unknown`.

The core issue here is that either the types are wrong (and `selectedOption` is really a `number`) or in this branch (`else`) the type of `selectedOption` is `string`, in which case the code is lying to typescript, which fails to discover that we're using a `string` where a `number` is expected, which would result in the wrong type of data being sent to the server. In that case, using `parseInt(selectedOption as string)` would have been a more correct thing to do.

When I look in the browser network panel to see what's actually getting sent to the server, I see a `number` here, so my analysis is that he types are wrong. The correct type of the `selectedOption` should probably be `number | SURVEY_QUESTION_OPTION`.

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+          },
+          title: selectedElement.question.question,
+          type: COLUMN_TYPE.SURVEY_RESPONSE,
+        },
+      ];
+    }
+  };
+
+  const makeColumnsOption = (
+    selectedOption: string | SURVEY_QUESTION_OPTIONS
```
Comment (richardolsson) [inline src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:68]: I'm now thinking this should be typed as `number | SURVEY_QUESTION_OPTIONS`.

File: src/features/views/components/ViewColumnDialog/SurveyResponsePluralConfig.tsx [hunk 6]
```diff
@@ -76,32 +76,35 @@ const SurveyResponsesConfig = ({
...
-                  }}
-                  label={intl.formatMessage({
-                    id: 'misc.views.columnDialog.choices.surveyResponses.questionField',
-                  })}
-                  variant="standard"
-                ></TextField>
-              )}
-              value={selectedQuestions}
-            />
+            {surveyId ? (
```
Comment (richardolsson) [inline src/features/views/components/ViewColumnDialog/SurveyResponsePluralConfig.tsx:79]: Like @ziggabyte suggested, this could be `{!!surveyId && (`

File: src/locale/misc/views/en.yml [hunk 7]
```diff
@@ -37,6 +37,18 @@ columnDialog:
       description: Add your choice of fields like email, phone number, etc.
       keywords: field multiple
       title: Person fields
+    surveyResponse:
+      allOptions: All Options (Single Column)
+      allOptionsHeader: All Options
+      allOptionsSeparated: All Options (Separate Columns)
+      description: Include response to a question from any survey.
+      keywords: field multiple
```
Comment (richardolsson) [inline src/locale/misc/views/en.yml:45]: This will need some more work, but that's what we'll be doing in the separate issue that @ziggabyte is creating, so no need to do anything about it here.

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+const SurveyResponseConfig = ({
+  onOutputConfigured,
+}: SurveyResponseConfigProps) => {
+  const { orgId } = useRouter().query;
+  const surveysQuery = useQuery(
+    ['surveysWithElements', orgId],
+    getSurveysWithElements(orgId as string)
+  );
+  const intl = useIntl();
+  const columnsReset: SelectedViewColumn[] = [];
```
Comment (richardolsson) [inline src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:42]: This variable seems unnecessary. It's only used in one place, and the literal `[]` could be used there instead.

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+                disabled={!surveyId}
+                fullWidth
+                getOptionLabel={(option) => option.question.question}
+                onChange={(evt, value) => {
+                  if (value !== null) {
+                    setSelectedQuestion(value);
+                    const columns = makeColumnsText(value);
+                    if (columns !== undefined) {
+                      onOutputConfigured(columns);
+                    }
```
Comment (richardolsson) [inline src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:156]: I find this logic to be difficult to follow. Unless I'm mistaken, what happens here is that `makeColumnsText()` is invoked for any question, even if it's not a text question.

For non-text questions, the function will return `undefined`, so on the next line we check for undefined before proceeding.

I think we should make sure that it's a text question here, before even calling `makeColumnsText()` and I think `makeColumnsText()` should only accept text questions (if there is a type for them, otherwise let's ignore this.).

Comment (richardolsson) [reply src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:156]: Looking into the types more closely, I see that there is no special type for text based questions, so maybe this can be ignored for now. Or we can refactor it together.

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+                  ></TextField>
+                )}
+                value={selectedQuestion}
+              />
+            ) : (
+              ''
+            )}
+            {surveyId &&
+            selectedQuestion &&
+            selectedQuestion?.question.response_type === RESPONSE_TYPE.TEXT ? (
```
Comment (richardolsson) [inline src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:179]: I don't like complicated ternaries like this in JSX. They are hard to follow, and it's difficult to see where they start and end. It's made even more difficult by the _nested_ ternary here (branches again on line 183).

If the logic isn't really nested, but just a way to make a three-way conditional, I would use the `condition && <Component/>` pattern three times instead.

**Review by richardolsson — CHANGES_REQUESTED** [review]
This works great and the logic is sound, but the code is difficult to follow and there are some unnecessary complexities that I think could be simplified. Let's chat if you want to pair program!

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx (hunk 5 above)
Comment (richardolsson) [reply src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:66]: For future reference: This was incorrect, but we solved it another way. 😊

File: src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx (hunk 1 above)
Comment (richardolsson) [reply src/features/views/components/ViewColumnDialog/SurveyResponseConfig.tsx:31]: This was never resolved.

**Review by richardolsson — APPROVED** [review]
We made it! Feel free to merge, @rebecarubio. 🎉 😎
